src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/Metadata/ModelBindingMessageProvider.cs

```diff
@@ -8,62 +8,62 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding.Metadata
-        Func<string> MissingKeyOrValueAccessor { get; }
+        public virtual Func<string> MissingKeyOrValueAccessor { get; }
 
         /// <summary>
         /// Error message the model binding system adds when no value is provided for the request body,
         /// but a value is required.
         /// </summary>
         /// <value>Default <see cref="string"/> is "A non-empty request body is required.".</value>
-        Func<string> MissingRequestBodyRequiredValueAccessor { get; }
+        public virtual Func<string> MissingRequestBodyRequiredValueAccessor { get; }
 
         /// <summary>
         /// Error message the model binding system adds when a <c>null</c> value is bound to a
         /// non-<see cref="Nullable"/> property.
         /// </summary>
         /// <value>Default <see cref="string"/> is "The value '{0}' is invalid.".</value>
-        Func<string, string> ValueMustNotBeNullAccessor { get; }
+        public virtual Func<string, string> ValueMustNotBeNullAccessor { get; }
 
         /// <summary>
         /// Error message the model binding system adds when <see cref="ModelError.Exception"/> is of type
         /// <see cref="FormatException"/> or <see cref="OverflowException"/> and value is known.
         /// </summary>
         /// <value>Default <see cref="string"/> is "The value '{0}' is not valid for {1}.".</value>
-        Func<string, string, string> AttemptedValueIsInvalidAccessor { get; }
+        public virtual Func<string, string, string> AttemptedValueIsInvalidAccessor { get; }
 
         /// <summary>
         /// Error message the model binding system adds when <see cref="ModelError.Exception"/> is of type
         /// <see cref="FormatException"/> or <see cref="OverflowException"/> and value is unknown.
         /// </summary>
         /// <value>Default <see cref="string"/> is "The supplied value is invalid for {0}.".</value>
-        Func<string, string> UnknownValueIsInvalidAccessor { get; }
+        public virtual Func<string, string> UnknownValueIsInvalidAccessor { get; }
 
         /// <summary>
         /// Fallback error message HTML and tag helpers display when a property is invalid but the
         /// <see cref="ModelError"/>s have <c>null</c> <see cref="ModelError.ErrorMessage"/>s.
         /// </summary>
         /// <value>Default <see cref="string"/> is "The value '{0}' is invalid.".</value>
-        Func<string, string> ValueIsInvalidAccessor { get; }
+        public virtual Func<string, string> ValueIsInvalidAccessor { get; }
 
         /// <summary>
         /// Error message HTML and tag helpers add for client-side validation of numeric formats. Visible in the
         /// browser if the field for a <c>float</c> property (for example) does not have a correctly-formatted value.
         /// </summary>
         /// <value>Default <see cref="string"/> is "The field {0} must be a number.".</value>
-        Func<string, string> ValueMustBeANumberAccessor { get; }
```

Comment:
Thanks, that explains the significance of the original design. I guess I'd side with keeping the `SetProperty(...)` methods and leaving the base class as immutable then (well, immutable as long as you don't cast to `DefaultModelBindingMessageProvider`, because then you can mutate it).